### ISSUE TITLE
RimWorld.FoodUtility.AddIngestThoughtsFromIngredient

### ISSUE THREAD

**Issue by Scorpiopt** [sumghai/Replimat#60]
**Describe the bug:**  
Since i built replimat items keep getting erros about job drivers and many of the pawns dont eat  , even if meals are still available from storage 

**Expected behavior:**  
Pawns to eat from Replimat or meals still left in storage 

**Steps to reproduce:**  
0 clue what started this 

**Additional information:**
Screenshots and Logs provided 

## System and Game Configuration
Processor: AMD Ryzen 5 3600X
Memory: 32GB Corsair Vengeance LPX 
Storage : - 1TB NVMe WD Blue SN550 
- 1TB SSD WD Blue 3D Nand
- 2TB SSD Samsung EVO 860 QVO
- 4TB Seagate BarraCuda 64 MB Cache SATA 7200 RPM
Video Card: RTX 2070 Super Windforce OC
OS: Windows 11 64bits 
**RimWorld base game version:** 1.3.3200
**RimWorld DLCs loaded:** Ideology and Royalty
**Replimat mod version:** Version 1.2.14 
**Replimat mod downloaded from:**  Steam 
![Untitled](https://user-images.githubusercontent.com/81530403/147798111-7c5da66f-b6fa-4184-afe8-8788c04a23c2.png)
![12222](https://user-images.githubusercontent.com/81530403/147798115-18d40cd6-1632-4ecb-aadf-390ca569dff8.png)

**Your mod list:** 
Link to mods list: https://rentry.co/c9c6z 
## Declaration

I hereby verify that:
- [X ] I am running the latest version of the game, Replimat and its dependencies
- [X ] I am NOT running any mods with known incompatibilities with Replimat 
- [ X] I have narrowed down my modlist to the fewest mods needed to cause this issue
- [ X] I have confirmed that the issue does not happen if Replimat is disabled
- [X ] I have searched the issue tracker and confirmed no-one else has reported my issue before
- [ X] I understand that my issue will be closed and ignored if I do not fill this form properly
[Player.zip](https://github.com/sumghai/Replimat/files/7795094/Player.zip)

**Comment by sumghai** [comment]
You have 579 mods in your mod list, any one of which may be interfering with Replimat's job drivers. Please conduct a binary search and narrow down your mod list to the bare minimum set of mods required to cause this issue, and update your initial report.

Also, without reproduction steps, there is no way I can recreate your problem on my end in order to diagnose what the problem is.

**Comment by Scorpiopt** [comment]
been doing that and checking whats patching the job drivers but nothing comes up 
RimmSQOL
Ben Lubars Vanilla Expanded patches
Performance Optimizer
Dragonian prime 2.0
Graze Up
Common Sense 
Alpha Animals
Vanilla expanded Social

Only thing left is the new Harmony update :?


JobDriver threw exception in initAction for pawn Arlene driver=JobDriver_Ingest (toilIndex=13) driver.job=(Ingest (Job_29102685) A=Thing_VCE_Gunkanmaki7753438 B=(205, 0, 163))
System.NullReferenceException: Object reference not set to an instance of an object
  at VanillaCookingExpanded.VanillaExpandedFramework_FoodUtility_GetMeatSourceCategory_Patch.DontCrapTheBedWithIngredientsWithoutNutrition (Verse.ThingDef source) [0x00001] in <527c4a8479df4472839011abb50877b2>:0 
  at (wrapper dynamic-method) RimWorld.FoodUtility.RimWorld.FoodUtility.GetMeatSourceCategory_Patch1(Verse.ThingDef)
  at RimWorld.FoodUtility.IsHumanlikeCorpseOrHumanlikeMeat (Verse.Thing source, Verse.ThingDef foodDef) [0x00017] in <cdbd0ed5089a418da09b9a259f9dbd8f>:0 
  at (wrapper dynamic-method) RimWorld.FoodUtility.RimWorld.FoodUtility.IsHumanlikeCorpseOrHumanlikeMeatOrIngredient_Patch0(Verse.Thing)
  at (wrapper dynamic-method) AlienRace.HarmonyPatches.AlienRace.HarmonyPatches.IngestedPrefix_Patch0(Verse.Pawn,Verse.Thing)
  at (wrapper dynamic-method) Verse.Thing.Verse.Thing.Ingested_Patch7(Verse.Thing,Verse.Pawn,single)
  at RimWorld.Toils_Ingest+<>c__DisplayClass12_0.<FinalizeIngest>b__0 () [0x00182] in <cdbd0ed5089a418da09b9a259f9dbd8f>:0 
  at Verse.AI.JobDriver.TryActuallyStartNextToil () [0x001de] in <cdbd0ed5089a418da09b9a259f9dbd8f>:0 
UnityEngine.StackTraceUtility:ExtractStackTrace ()
(wrapper dynamic-method) Verse.Log:Verse.Log.Error_Patch3 (string)
Verse.AI.JobUtility:TryStartErrorRecoverJob (Verse.Pawn,string,System.Exception,Verse.AI.JobDriver)
Verse.AI.JobDriver:TryActuallyStartNextToil ()
Verse.AI.JobDriver:ReadyForNextToil ()
(wrapper dynamic-method) Verse.AI.JobDriver:Verse.AI.JobDriver.DriverTick_Patch0 (Verse.AI.JobDriver)
(wrapper dynamic-method) Verse.AI.Pawn_JobTracker:Verse.AI.Pawn_JobTracker.JobTrackerTick_Patch1 (Verse.AI.Pawn_JobTracker)
(wrapper dynamic-method) Verse.Pawn:Verse.Pawn.Tick_Patch4 (Verse.Pawn)
Verse.TickList:Tick ()
(wrapper dynamic-method) Verse.TickManager:Verse.TickManager.DoSingleTick_Patch4 (Verse.TickManager)
Verse.TickManager:TickManagerUpdate ()
(wrapper dynamic-method) Verse.Game:Verse.Game.UpdatePlay_Patch1 (Verse.Game)
Verse.Root_Play:Update ()

**Comment by sumghai** [comment]
The latest (31 Dec 2021) Harmony update hasn't caused any issues, and neither does the latest versions of Vanilla Expanded (271221) and Humanoid Alien Races (4 Dec 2021) Framework. I started a new savegame with Fungus preferred precept active, and pawns replicated meals just fine. I also had no problems with existing non-Ideology savegames.

Please keep trying to narrow down your modlist, until you find the conflicting mod(s).

**Comment by Scorpiopt** [comment]
Ok i moved Replimat Before Vanilla expanded Framework since all errors are related to VE Cooking meals , and so far error is not showing up again.

My best guess is that their transpiler is breaking the one from Replimat . 
Same way VE Framework's transpiler breaks Ddubs aparel tweaks transpiler

**Comment by Scorpiopt** [comment]
moved replimat before VEF in hopes it was their transpiller breaking it since its their meals only causing the issue but nope it was only temporary before it returned

**Comment by Scorpiopt** [comment]
cause of the problem found  https://steamcommunity.com/sharedfiles/filedetails/?id=2559533848

Sadly too late to save my game because the damage is done

**Comment by sumghai** [comment]
While I don't use the More Precepts mod myself, it looks like it is trying to transpile `FoodUtility.BestFoodSourceOnMap`, whereas Replimat simply prefixes and postfixes the same method.

I may have to get in touch with the mod author to see if we can figure out something.

> Sadly too late to save my game because the damage is done , even removal wont fix it

That's bizarre. Surely removing either Replimat or More Precepts would mean only the remaining mod is patching `BestFoodSourceOnMap`, and therefore no longer cause any conflicts? Unless, you have *yet another* mod that's *also* trying to patch that method?

**Comment by Scorpiopt** [comment]
Looked into every patch every stacktrace , maybe some leftover code screwed up the save ,more precepts has definitely been removed ( even edited ideology to remove the rituals i was using from it and precepts ) and so has replimat after destroying every item of it on the map

**Comment by Scorpiopt** [comment]
upon checking the new logs for errors can confirm more precepts does leave traces on the save file causing issues and save editing to remove it beyond my capabilities

**Comment by sumghai** [comment]
I did some testing on my end just now, and I was unable to reproduce the issue at all with Replimat 1.2.14, More Precepts (1 Sept 2021), Vanilla Expanded Framework (27 Dec 2021) and Humanoid Alien Races (4 Dec 2021), as well as the latest versions of Royalty and Ideology, with the Alcohol possession precept from More Precepts.

I'll also note that at game startup, the debug log displayed an error stating `MorePrecepts: Failed to find BestFoodSourceOnMap_foodValidator for patching`. This error persisted even if I removed every other mod (including Replimat).

It has been reported by many users that More Precepts itself, in fact, is no longer patching the food optimality method properly since the most recent RimWorld update. Vanilla Expanded Framework was even updated in an attempt to work around the More Precepts bug, to no avail:

![Screenshot 2022-01-02 085753](https://user-images.githubusercontent.com/4993380/147863436-df0179a4-75fe-48f6-82a4-56cf27bffc7c.png)

While I understand your frustration with losing your savegame, given these facts, there is nothing I can do on my end. The problem most likely lies solely with More Precepts itself, so you'll have to wait until llunak comes back to fix their mod.
